plugins/datadog/datadog.go

```diff
@@ -0,0 +1,185 @@
+	}
+}
+
+func main() {
+	log.Info("Starting datadog DC/OS metrics plugin")
+
+	ddPluginFlags := []cli.Flag{
+		cli.StringFlag{
+			Name:  "datadog-host",
+			Value: "localhost",
+			Usage: "Datadog URL to query",
+		},
+		cli.IntFlag{
+			Name:  "datadog-port",
+			Value: 8125,
+			Usage: "Datadog port",
+		},
+	}
+
+	ddp, err := plugin.New(ddPluginFlags)
+	handleErr(err)
+
+	ddp.Action = func(c *cli.Context) error {
+		datadogAgent := fmt.Sprintf("%s:%d", c.String("datadog-host"), c.Int("datadog-port"))
+
+		metricsPaths, err := plugin.MetricsEndpoints(c.String("dcos-role"))
+		if err != nil {
+			return err
+		}
+
+		for {
+			for _, path := range metricsPaths {
+				metricsURL := url.URL{
+					Scheme: c.String("metrics-proto"),
+					Host:   fmt.Sprintf("%s:%d", c.String("metrics-host"), c.Int("metrics-port")),
+					Path:   path,
+				}
+				token := c.String("auth-token")
+				if len(token) == 0 {
+					return errors.New("Auth token must be set, use --auth-token <token>")
+				}
+				request := &http.Request{
+					Method: "GET",
+					URL:    &metricsURL,
+					Header: http.Header{
+						"Authorization": []string{fmt.Sprintf("token=%s", token)},
+					},
+				}
+
+				metricMessage, err := getDCOSMetrics(request)
+				if err != nil {
+					return err
+				}
+
+				log.Infof("Received data from metrics service endpoint %s, success!", request.URL.Path)
+
+				if err := toStats(metricMessage, datadogAgent); err != nil {
+					log.Errorf("Errors encountered sending metrics to Datadog: %s", err.Error())
+					continue
+				}
+
+			}
+			interval := c.Int("polling-interval")
+			log.Infof("Polling complete, sleeping for %d seconds", interval)
+			time.Sleep(time.Duration(interval) * time.Second)
+		}
+		return nil
+	}
+
+	ddp.Run(os.Args)
+}
+
+func toStats(mm producers.MetricsMessage, ddURL string) error {
+	c, err := statsd.New(ddURL)
+	if err != nil {
+		return err
+	}
+
+	c.Namespace = "dcos-metrics"
+
+	tags, err := buildTags(mm.Dimensions)
+	if err != nil {
+		return err
+	}
+
+	for _, dp := range mm.Datapoints {
+		name := strings.Join([]string{mm.Name, dp.Name}, producers.MetricNamespaceSep)
+		val, err := strconv.ParseFloat(fmt.Sprintf("%v", reflect.ValueOf(dp.Value).Interface()), 64)
+		if err != nil {
+			return err
+		}
+
+		err = c.Gauge(name, val, tags, 1)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// buildTags analyzes the MetricsMessage.Dimensions struct and returns a slice
+// of strings of key/value pairs in the format "key:value"
+func buildTags(msg producers.Dimensions) (tags []string, err error) {
+	v := reflect.ValueOf(msg)
+	if v.Kind() != reflect.Struct {
+		return tags, fmt.Errorf("error: expected type struct, got %s", v.Kind().String())
+	}
+
+	for i := 0; i < v.NumField(); i++ {
+		// The producers.Dimensions struct contains a nested
+		// map[string]string for user-defined Labels
+		if v.Field(i).Kind() == reflect.Map {
+			for k, v := range v.Field(i).Interface().(map[string]string) {
```

Comment:
while you check `v.Field(i).Kind() == reflect.Map` you type assert to a map[string]string. If this happens to be a `map[string]interface{}` for some reason, this will panic in runtime.